doc/decisions/array.md

```diff
@@ -40,11 +38,30 @@ myarray/#5 = value5
   array = #5
 ```
 
+To lookup an array, first do `ksLookup(ks,"/myarray")` on the parent.
+With the last index you get from its metadata `array`, iterate over the children.
```

Comment:
Using the `array` metadata on the parent can actually cause inconsistent data. Only one file, i.e. one mountpoint, can be read atomically. But the parent key could be in a different mountpoint (a weird setup, but possible)

Reply:
I would not call it inconsistent, it might only wrongly turn from array to non-array.

Reply:
What would you call inconsistent, if not a completely wrong interpretation of the config?

Just think about `spec`. The `array` metadata makes a huge difference to what happens and might even cause some errors.

Reply:
Of course it is inconsistent on the level of configuration but this it would be either way if one mountpoint gets written and the other does not. Btw. in the case of a computer crash the non-renamed files would still be lying there, so the admin can make the configuration consistent again.

But what I meant is that it is consistent from the point of view of the array:

- if the array metadata is there, it is an array
- if the array metadata is not there, it is not an array

What we did not yet decide is what should happen with an array where the children have keys which are not in `#...` syntax. Basically, we could normalize it during `kdbGet` (if we decide to do something like #3497).

Reply:
> Of course it is inconsistent on the level of configuration but this it would be either way if one mountpoint gets written and the other does not. Btw. in the case of a computer crash the non-renamed files would still be lying there, so the admin can make the configuration consistent again.

I wasn't talking about outright crashes. I was just thinking about a race condition. There could be two processes accessing the KDB. Process A wants to update both mountpoints, but process B reads them when only one of the two mountpoints has been updated.

> But what I meant is that it is consistent from the point of view of the array:

What if the array size was changed. There could be missing array elements now, which might cause unexpected errors.

> What we did not yet decide is what should happen with an array where the children have keys which are not in #... syntax. Basically, we could normalize it during kdbGet

What do you mean by "normalize"?

Reply:
> Process A wants to update both mountpoints, but process B reads them when only one of the two mountpoints has been updated.

There are mechanisms in resolver to prevent this (checking if the file was tempted with before renaming). And Elektra also locks, so if both processes use Elektra they should be safe from this.

> What if the array size was changed. There could be missing array elements now, which might cause unexpected errors.

All of this should be caught by `spec` in kdbGet.

> What do you mean by "normalize"?

The `spec` plugin could even rewrite:
```
user:/myarray
user:/myarray/a
user:/myarray/b
user:/myarray/c
```

to:

```
user:/myarray      # array=#2
user:/myarray/#0
user:/myarray/#1
user:/myarray/#2
```

Reply:
> The spec plugin could even rewrite: [...]

Are you serious?? That's so much black magic that even sys-admin-wizards would be confused for days why the application receives a config completely different from the one in the file....

Reply:
> And Elektra also locks, so if both processes use Elektra they should be safe from this.

Yes, but [afaik](http://0pointer.de/blog/projects/locking.html) this can not be done reliably and portably.

Reply:
> Are you serious??

Actually it is current behavior for some storage plugins, e.g. fstab iirc.

> That's so much black magic that even sys-admin-wizards would be confused for days why the application receives a config completely different from the one in the file....

In the file is always the correct (it always contains an array), it is only about:

```
kdb set user:/myarray/zzz sthg
kdb get user:/myarray/zzz
#RET: 2
#> key not found
kdb get user:/myarray/#3
#> sthg
```

So actually it is currently a workaround for the missing array functionality in `kdb` (for appending to arrays).

I don't say I like it, but it is also not soo bad and quite practical (at least as long we do not have proper tooling support for handling arrays).

Reply:
> Actually it is current behavior for some storage plugins, e.g. fstab iirc.

Yes, a specialised storage plugin like `fstab` could do such things, but not `spec`. If the user chooses `fstab` they explicitly opt in to such behaviour, but `spec` should be the default.

> The file is always the correct (it always contains an array), it is "only" about: [...] not soo bad and quite practical

I'm sorry, I cannot accept that. If that is actually your opinion, I think it would best for me to stop working on Elektra, because we will never agree on anything.

That

```
kdb set user:/myarray/zzz sthg
kdb get user:/myarray/zzz
#RET: 2
#> key not found
```

should **NEVER EVER EVER** happen. That is the **WORST BUG** that Elektra could possible have. It failed at its single most important job: Storing a value for a key.

Reply:
:laughing: Of course it is a bug: the validation is missing. But I would not call it the worst bug, e.g. missing keys in the multifile resolver are definitely worse (`kdb set` which goes to nirvana.)

Reply:
IMO `kdb set` changing a different key than stated without any indication that this happend is worse than it simply doing nothing. In your example the key `user:/myarray/#3` may have contained important stuff that is now gone, because `kdb set` did something very unexpected.

It's like saying:

It's not so bad that `echo "test" > /etc/useless_file` wrote to `/etc/passwd` instead, at least it didn't do nothing.

Reply:
:laughing: Of course this is impossible, if `#3` would exist, it would have been called `#4`. But your whole argumentation basically points that you actually do **not** expect it to be an array but rather a convention of names. Otherwise you would not be **so** repulsive to such ideas, e.g. Python also uses out-of-bound index for something else (e.g. -1 for the last entry). But I agree that all of this is in the best case a workaround, what we actually want is a nice interface to add entries to arrays.

Reply:
> Of course this is impossible, if `#3` would exist, it would have been called `#4`

That doesn't seem much better. Okay, we don't loose values, but we still do something quite unexpected.

> But your whole argumentation basically points that you actually do not expect it to be an array but rather a convention of names. Otherwise you would not be so repulsive to such ideas, e.g. Python also uses out-of-bound index for something else (e.g. -1 for the last entry).

No, I actually want it to be arrays. The Python thing is different, `-1` is still a number and Python also doesn't accept all numbers as indices for every array. Only `n,...,n-1` are allowed for an array of length `n`. You can only read from or write to these indices. The only difference between Python and e.g. Java is that each array element has 2 indices (on positive and on negative).

> But I agree that all of this is in the best case a workaround, what we actually want is a nice interface to add entries to arrays.

Good then, we _do_ agree. :)